%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Professional Formal Letter
% LaTeX Template
% Version 1.0 (28/12/13)
%
% This template has been downloaded from:
% http://www.LaTeXTemplates.com
%
% Original author:
% Brian Moses (http://www.ms.uky.edu/~math/Resources/Templates/LaTeX/)
% with extensive modifications by Vel (vel@latextemplates.com)
%
% License:
% CC BY-NC-SA 3.0 (http://creativecommons.org/licenses/by-nc-sa/3.0/)
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%----------------------------------------------------------------------------------------
%	PACKAGES AND OTHER DOCUMENT CONFIGURATIONS
%----------------------------------------------------------------------------------------

\documentclass[11pt,letterpaper]{letter} % Specify the font size (10pt, 11pt and 12pt) and paper size (letterpaper, a4paper, etc)

\usepackage{umass_letterhead}

\begin{document}

\vspace{\fill}
\begin{letter}{}

\opening{Dear Editors,}

We are submitting our manuscript ``Variational inference for rare variant detection in deep, heterogeneous next-generation sequencing data'' for your consideration as an original methodology article to \textit{BMC Bioinformatics}. 
We developed a variational expectation maximization (EM) inference algorithm to detect rare single nucleotide variants and estimate non-reference allele frequencies in heterogeneous samples.
We show the accuracy of our variational EM algorithm is better than many state-of-the-art algorithms. 
We detected rare variants with low non-reference allele frequencies (<1.0\%) early in the time course of a \textit{S. cerevisiae} directed evolution experiment data set.

We believe \textit{BMC Bioinformatics} is the best venue for our article because it is open-access and publishes high-quality research on statistical methods for modeling next-generation sequencing data.
Our article fits well with these aims and scope.

\closing{Sincerely,}

\vspace{\fill}
\end{letter}

\end{document}
